Add clustering documentation (#122)
aguirre-lab/ml4c3-wiki: Clustering.md

# Clustering

> Clustering is a mode that lets you find a cohort, extract data from that cohort

Reviewer: ```suggestion
Clustering is a recipe that lets a user define a cohort, extract data from that cohort,
```

and perform operations on it, like QC, plotting or clustering.

The clustering code is thought to be used as an imported dependency on a python 

> shell. It additionally contains two recipies that can be called from the terminal:

Reviewer: ```suggestion
shell. It additionally contains two recipes that can be called from the terminal:
```

`Find` and `Extract`.

> ## Find
> 
> #### From the terminal 
> The find recipe will look for files containing the input signals if they are recoreded
> simultaneously during the BLK08 stay of the patient.

Reviewer: ```suggestion
## `finder`

#### From the terminal 
The `finder` recipe will look for files containing the input signals if they are recorded
simultaneously during the BLK08 stay of the patient.
```

Run the find mode with the command:
```shell script
./scripts/run.sh $PWD/ml4c3/recipes.py cluster finder \
  --tensors /media/mad3/hd5/ \

>   --signals art1d art1m art1s pa2s pa2d pa2m hr \

Reviewer: ```suggestion
  --signals art1d art1m art1s pa2s pa2d pa2m hr \
```

We will need to replace this with tensor maps in future PR

|  6 |   1319 |   1319 |   1319 |    1319 |    1319 |    1319 | 1319 | 5166848.hd5 | 3139183944 |                 24 |                    1319 |                            0 |                          0 | True      |              23.8108 |               99.2118 |
|  7 |   1197 |   1197 |   1197 |    1197 |    1197 |    1197 | 1197 | 2053794.hd5 | 3128446495 |                 24 |                    1197 |                            0 |                          0 | True      |              23.9997 |               99.9988 |
|  8 |    577 |    577 |    577 |     577 |     577 |     577 |  577 | 2147687.hd5 | 3138253743 |                 24 |                     577 |                            0 |                          0 | True      |              23.8386 |               99.3275 |
|  9 |     15 |     15 |     15 |     254 |     254 |     254 |   15 | 1402793.hd5 | 3115818304 |                 24 |                     254 |                            0 |                          0 | True      |              23.9294 |               99.706  |

> 

Reviewer: Please confirm these MRNs and CSNs are fake.

the period studied.
* `Overlap percent (%)`: The percentage of time that the signals are recoreded 
simultaneously over the period studied.

##### Cell value
Note that column statistics represent the maximum values between all signals.
If you want to see a statistic for a specific signal, you can do it with the 
`--cell_value` parameter. On the example above, `cell_value` is `max_unfilled` so on
the columns of each signal is displayed the maximum unfilled time specific for each 
signal, while the column `Max unfilled time (s)` contains the maximum between all the 
signals. 

Valid cell values are:
* existence: boolean indicating if the signal is found on the file and visit
* overlap_length: duration of the signal inside the period limits
* overlap_percent: duration of the signal with respect to the duration of the period 
limits (%)
* start_time: timestamp for the beginning of the signal inside of the period studied
* end_time: timestamp for the ending of the signal inside of the period studied
* max_unfilled: maximum gap inside the signal without values (in seconds)
* non_monotonicities: number of non-monotonicities on the signal

> * max_non_monotonicity: maximum non-monotonicity on the signal

Reviewer: Can you quote format the variable names?

#### On Python

> You can use run `Explore` class on a python shell or script:

Reviewer: ```suggestion
You can use the `Explorer` class in Python, e.g. in a notebook:
```

```python
from clustering.objects.finder import Explorer
signals = ["hr", "art1m", "pa2m"]
explorer = Explorer("some/path")
report = explorer.find(signals, cell="max_unfilled", stay_length=12)
``` 
Or if you have generated the csv from the terminal you can read it running:
```python
from clustering.objects.finder import RequestReport
report = RequestReport.from_csv("path/to/csv")
``` 

> The `Explorer` and `RequestReport` class have useful features like:
> 
> * Searching all the signals on a path that contain a particular keyword. NOOTEBOOK LINK
> * Plotting all the signals on a file.
> * Plotting a signal across all the files in a directory that contain it.

Reviewer: ```suggestion
The `Explorer` and `RequestReport` classes can

* Search all the signals on a path that contain a particular keyword. NOTEBOOK LINK
* Plot all the signals on a file.
* Plot a signal across all the files in a directory that contain it.
```

There are many more features. Check:

> * The [Github code](https://github.com/aguirre-lab/ml4c3/blob/d501b4db027e597123dd3ccb8a1f73a7d97aac01/clustering/objects/finder.py#L18-L297) 

Reviewer: ```suggestion
* [code](https://github.com/aguirre-lab/ml4c3/blob/d501b4db027e597123dd3ccb8a1f73a7d97aac01/clustering/objects/finder.py#L18-L297) 
```

> * The [Example notebook](https://github.com/aguirre-lab/ml4c3/blob/master/notebooks/clustering/clustering_example.ipynb)

Reviewer: ```suggestion
* [demo notebook](https://github.com/aguirre-lab/ml4c3/blob/master/notebooks/clustering/clustering_example.ipynb)
```

> ## Extract
> 
> Extract recipie uses a `.csv` in the same format than the output from the 

Reviewer: ```suggestion
## `extract`

This recipe uses a `.csv` in the same format than the output from the 
```

> [Find command](#find) and extracts all the signals, storing them in a 

Reviewer: ```suggestion
[`finder`](#finder) and extracts all the signals, storing them in a 
```

[Bundle](#bundle) class. 

The normal workflow is to filter out the csv generated by `Find` based on some 
conditions and use the filtered csv as input for `Extract`, in order to get a 
cohort that meets certain criteria.

#### From the terminal

> You can run this mode with:

Reviewer: ```suggestion
You can run this recipe with:
```
